File: go-client/internal/network/discovery.go
package network

import (
	"context"
	"fmt"
	"log"
	"net"
	"sync"
	"time" // used by DiscoveredPeer.LastSeen

	"github.com/grandcat/zeroconf"
)

const serviceType = "_p2pshare._tcp"
const domain = "local."

// DiscoveredPeer is a peer found via mDNS.
type DiscoveredPeer struct {
	Name     string
	PeerID   string
	Host     string
	Port     int
	LastSeen time.Time
}

// PeerDiscovery manages mDNS service registration and browsing.
type PeerDiscovery struct {
	DisplayName string
	PeerID      string
	Port        int

	mu     sync.Mutex
	peers  map[string]*DiscoveredPeer // keyed by peer_id
	server *zeroconf.Server
	cancel context.CancelFunc
}

// NewPeerDiscovery creates a new PeerDiscovery instance.
func NewPeerDiscovery(displayName, peerID string, port int) *PeerDiscovery {
	return &PeerDiscovery{
		DisplayName: displayName,
		PeerID:      peerID,
		Port:        port,
		peers:       make(map[string]*DiscoveredPeer),
	}
}

// Start begins mDNS registration and browsing.
func (pd *PeerDiscovery) Start() error {
	// Register our service
	var err error
	pd.server, err = zeroconf.Register(
		pd.DisplayName,  // instance name
		serviceType,     // service type
		domain,          // domain
		pd.Port,         // port
		[]string{fmt.Sprintf("peer_id=%s", pd.PeerID)}, // TXT records
		nil,             // interfaces (nil = all)
	)
	if err != nil {
		return fmt.Errorf("mDNS register: %w", err)
	}

	// Browse for peers
	ctx, cancel := context.WithCancel(context.Background())
	pd.cancel = cancel
	go pd.browse(ctx)

	return nil
}

// Stop stops mDNS registration and browsing.
func (pd *PeerDiscovery) Stop() {
	if pd.cancel != nil {
		pd.cancel()
	}
	if pd.server != nil {
		pd.server.Shutdown()
	}
}

// GetPeers returns all discovered peers (excluding self).
func (pd *PeerDiscovery) GetPeers() []*DiscoveredPeer {
	pd.mu.Lock()
	defer pd.mu.Unlock()
	var result []*DiscoveredPeer
	for _, p := range pd.peers {
		if p.PeerID != pd.PeerID {
			result = append(result, p)
		}
	}
	return result
}

// GetPeer looks up a discovered peer by peer_id.
func (pd *PeerDiscovery) GetPeer(peerID string) *DiscoveredPeer {
	pd.mu.Lock()
	defer pd.mu.Unlock()
	return pd.peers[peerID]
}

func (pd *PeerDiscovery) browse(ctx context.Context) {
	resolver, err := zeroconf.NewResolver(nil)
	if err != nil {
		log.Printf("mDNS resolver error: %v", err)
		return
	}

	entries := make(chan *zeroconf.ServiceEntry)
	go func() {
		for entry := range entries {
			pd.addPeer(entry)
		}
	}()

	// A single Browse runs until ctx is cancelled by Stop().
	// Do NOT restart Browse on the same entries channel — zeroconf closes it
	// on context cancellation and a second Browse would panic (close of closed channel).
	if err := resolver.Browse(ctx, serviceType, domain, entries); err != nil {
		log.Printf("mDNS browse error: %v", err)
	}
}

func (pd *PeerDiscovery) addPeer(entry *zeroconf.ServiceEntry) {
	var peerID string
	for _, txt := range entry.Text {
		if len(txt) > 8 && txt[:8] == "peer_id=" {
			peerID = txt[8:]
		}
	}
	if peerID == "" {
		return
	}

	host := ""
	if len(entry.AddrIPv4) > 0 {
		host = entry.AddrIPv4[0].String()
	} else if len(entry.AddrIPv6) > 0 {
		host = entry.AddrIPv6[0].String()
	}
	if host == "" {
		return
	}

	pd.mu.Lock()
	defer pd.mu.Unlock()
	pd.peers[peerID] = &DiscoveredPeer{
		Name:     entry.Instance,
		PeerID:   peerID,
		Host:     host,
		Port:     entry.Port,
		LastSeen: time.Now(),
	}
}

// GetLocalIP returns the machine's non-loopback IP address.
func GetLocalIP() string {
	addrs, err := net.InterfaceAddrs()
	if err != nil {
		return "127.0.0.1"
	}
	for _, addr := range addrs {
		if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
			if ipnet.IP.To4() != nil {
				return ipnet.IP.String()
			}
		}
	}
	return "127.0.0.1"
}

File: python-client/p2pshare/network/discovery.py
"""mDNS/DNS-SD peer discovery using zeroconf.

Service type: _p2pshare._tcp.local.
TXT records: peer_id=<base64url>, port=<tcp_port>
"""

import socket
import threading
import time

from zeroconf import Zeroconf, ServiceBrowser, ServiceInfo, ServiceStateChange


SERVICE_TYPE = "_p2pshare._tcp.local."


class DiscoveredPeer:
    """A peer discovered via mDNS."""

    def __init__(self, name: str, peer_id: str, host: str, port: int):
        self.name = name
        self.peer_id = peer_id
        self.host = host
        self.port = port
        self.last_seen = time.time()

    def __repr__(self):
        return f"DiscoveredPeer({self.name}, {self.host}:{self.port}, id={self.peer_id[:8]}...)"


class PeerDiscovery:
    """Manages mDNS service registration and browsing for peer discovery."""

    def __init__(self, display_name: str, peer_id: str, port: int):
        self.display_name = display_name
        self.peer_id = peer_id
        self.port = port
        self.peers: dict[str, DiscoveredPeer] = {}  # keyed by peer_id
        self._lock = threading.Lock()
        self._zeroconf: Zeroconf | None = None
        self._browser: ServiceBrowser | None = None
        self._service_info: ServiceInfo | None = None

    def start(self) -> None:
        """Start mDNS registration and browsing."""
        self._zeroconf = Zeroconf()

        # Register our service
        hostname = socket.gethostname()
        local_ip = self._get_local_ip()

        self._service_info = ServiceInfo(
            SERVICE_TYPE,
            f"{self.display_name}.{SERVICE_TYPE}",
            addresses=[socket.inet_aton(local_ip)],
            port=self.port,
            properties={
                "peer_id": self.peer_id.encode("utf-8"),
            },
            server=f"{hostname}.local.",
        )
        self._zeroconf.register_service(self._service_info)

        # Browse for peers
        self._browser = ServiceBrowser(
            self._zeroconf, SERVICE_TYPE, handlers=[self._on_service_state_change]
        )

    def stop(self) -> None:
        """Stop mDNS registration and browsing."""
        if self._zeroconf:
            if self._service_info:
                self._zeroconf.unregister_service(self._service_info)
            self._zeroconf.close()
            self._zeroconf = None

    def get_peers(self) -> list[DiscoveredPeer]:
        """Return a list of currently discovered peers (excluding self)."""
        with self._lock:
            return [p for p in self.peers.values() if p.peer_id != self.peer_id]

    def get_peer(self, peer_id: str) -> DiscoveredPeer | None:
        """Look up a discovered peer by peer_id."""
        with self._lock:
            return self.peers.get(peer_id)

    def _on_service_state_change(
        self, zeroconf: Zeroconf, service_type: str, name: str, state_change: ServiceStateChange
    ) -> None:
        if state_change == ServiceStateChange.Added:
            info = zeroconf.get_service_info(service_type, name)
            if info:
                self._add_peer(info)
        elif state_change == ServiceStateChange.Removed:
            self._remove_peer(name)

    def _add_peer(self, info: ServiceInfo) -> None:
        props = info.properties or {}
        peer_id_bytes = props.get(b"peer_id", b"")
        if isinstance(peer_id_bytes, bytes):
            peer_id = peer_id_bytes.decode("utf-8")
        else:
            peer_id = str(peer_id_bytes)

        if not peer_id:
            return

        addresses = info.parsed_addresses()
        if not addresses:
            return
        host = addresses[0]

        peer = DiscoveredPeer(
            name=info.name,
            peer_id=peer_id,
            host=host,
            port=info.port,
        )
        with self._lock:
            self.peers[peer_id] = peer

    def _remove_peer(self, name: str) -> None:
        with self._lock:
            to_remove = [pid for pid, p in self.peers.items() if p.name == name]
            for pid in to_remove:
                del self.peers[pid]

    def _get_local_ip(self) -> str:
        """Get the local IP address (not 127.0.0.1) for mDNS advertisement."""
        try:
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.connect(("8.8.8.8", 80))
            ip = s.getsockname()[0]
            s.close()
            return ip
        except Exception:
            return "127.0.0.1"
